File: behavioral/command-example.js
class CmdKill {

    constructor(name) {
        this.name = name;
    }

    run(...args) {
        console.log(`${this.name} is being killed at ${args[0]}!`);
    }
}

// ---------------------------------------------------------
class CmdBeat {
    constructor(name) {
        this.name = name
    }

    run(...args) {
        console.log(`${this.name} is being beaten at ${args[0][0]}, ${args[0][1]} times`)
    }
}

// ---------------------------------------------------------
class Invoker0 {

    constructor(cmd, ...args) {
        this.cmd = cmd;
        this.args = args;
    }

    invoke() {
        this.cmd.run(this.args);
    }
}

// ---------------------------------------------------------
class Invoker1 {

    static invoke(cmd, ...args) {
        cmd.run(args);
    }
}


// ---------------------------------------------------------
class Invoker2 {
    constructor() {
        this.cmd = null;
        this.args = null;
    }

    set(cmd, ...args) {
        this.cmd = cmd;
        this.args = args;
    }

    invoke() {
        this.cmd.run(this.args);
    }

}

// ---------------------------------------------------------
let c = new CmdKill('tina');
c.run('10:13');

console.log('- - - - - - - - - - - - - - - -');

let i1 = new Invoker1();
Invoker1.invoke(new CmdBeat('kamran'), '10:10', 2);
Invoker1.invoke(new CmdKill('kamran'), '10:13');

console.log('- - - - - - - - - - - - - - - -');

i1 = new Invoker2();
i1.set(new CmdBeat('kamran'), '10:10', 2);
i1.invoke();

i1.set(new CmdKill('kamran'), '10:13');
i1.invoke();

console.log('- - - - - - - - - - - - - - - -');

i1 = new Invoker0(new CmdBeat('kamran'), '10:10', 2);
i1.invoke();

i2 = new Invoker0(new CmdKill('kamran'), '10:13');
i2.invoke();

File: behavioral/command-example.py
import logging

logging.basicConfig(format='%(asctime)s %(levelname)-8s %(message)s', level=logging.INFO, datefmt='%Y-%m-%d %H:%M:%S')


class CmdKill:
    def __init__(self, name):
        self._name = name

    def run(self, *args):
        print(f'{self._name} is being killed at {args[0][0]}!')


class CmdBeat:
    def __init__(self, name):
        self._name = name

    def run(self, *args):
        print(f'{self._name} is being beaten at {args[0][0]}, {args[0][1]} times')


class Invoker0:
    def __init__(self, cmd, *args):
        self._cmd = cmd
        self._args = args

    def invoke(self):
        logging.warning(f'Invoking command {type(self._cmd).__name__} with parameters {self._args}')
        self._cmd.run(self._args)


class Invoker1:
    @staticmethod
    def invoke(cmd, *args):
        logging.warning(f'Invoking command {type(cmd).__name__} with parameters {args}')
        cmd.run(args)


class Invoker2:
    def __init__(self):
        self._cmd = None
        self._args = None

    def set(self, cmd, *args):
        self._cmd = cmd
        self._args = args

    def invoke(self):
        logging.warning(f'Invoking command {type(self._cmd).__name__} with parameters {self._args}')
        self._cmd.run(self._args)


if __name__ == "__main__":
    c = CmdKill('tina')
    c.run('10:13')

    print('- ' * 30)

    Invoker1.invoke(CmdBeat('kamran'), '10:10', 2)
    Invoker1.invoke(CmdKill('kamran'), '10:13')

    print('- ' * 30)

    i1 = Invoker2()
    i1.set(CmdBeat('kamran'), '10:10', 2)
    i1.invoke()

    i1.set(CmdKill('kamran'), '10:13')
    i1.invoke()

    print('- ' * 30)

    i1 = Invoker0(CmdBeat('kamran'), '10:10', 2)
    i1.invoke()

    i2 = Invoker0(CmdKill('kamran'), '10:13')
    i2.invoke()
